### DISCUSSION TITLE
Profilering av pagelinks

### DISCUSSION THREAD

**Discussion by TheodorRene** [TheodorRene/Prosjektoppgave#4] [General]
Bare for å fortsette tråden fra i går.  Ingen diskusjon om optimisering er verdt noe uten noe profilering og tall. Ish 95% av tiden scriptet bruker er på nettverkskall. Så å bytte programmeringspråk vil nok ikke endret noe. (Jeg pusher en commit med endringene mine, men de trenger ikke å bli med videre. Måtte gjøre timinglogikken manuelt for å gjøre dekoratoren stateful)

![Screenshot_2021-09-27_15-05-01](https://user-images.githubusercontent.com/36312737/134917592-f3c419be-406e-4ae3-9c81-9bb50cd85346.png)

Implementert her (https://github.com/TheodorRene/Prosjektoppgave/pull/1/commits/c79c222439813e005614713a08e77c695d382116)

**Comment by eivindkopperud** [comment]
Jepp, testet det og sjekket selv. Godt tenkt, og er helt enig. Jeg har sett litt mer på Wikipedia API-et og det er nok noen hensyn vi må ta.

Wikipedia er ganske liberal når det kommer til antall requests man sender; det er ingen rate limiting. Likevel nevner de at det er fint om man gjør API-kall med omhu. Vi kan batche alle interne linker for hver side i én request ganske lett ved å bare legge til flere sider i titles-parameteret.

Jeg tror også at det går helt fint an å komme unna å laste ned alt innhold med data for hver eneste request for å finne lenkene. Ved å legge til `ids` i rvprop, så kan vi få en revisjonsid for en gitt artikkel:

```
https://en.wikipedia.org/w/api.php?action=query&format=json&prop=revisions&rvlimit=20&titles=Erna_Solberg&rvlimit=max&rvprop=timestamp|ids
```

Deretter kan vi hente ut alle lenkene til disse revisionene vha. følgende:

```
https://en.wikipedia.org/w/api.php?action=query&revids=1046258009&prop=links&pllimit=max&format=json
```
Requesten over henter alle interne wiki-lenker til den nyeste [Erna Solberg-revisjonen](https://en.wikipedia.org/wiki/Erna_Solberg). Ser derimot ut som at det er limit på 500 lenker - har til gode for å finne ut hvordan pagination fungerer her.

**Reply by eivindkopperud** [reply]
**Pagination på API-kall**
Som nevnt i posten over, kan man maks hente ut 500 linker. For å hente ut de neste, bruker man plcontinue som man får i responsen.
```
https://en.wikipedia.org/w/api.php?action=query&revids=1046258009&prop=links&pllimit=max&format=json&plcontinue=651819|0|North_Macedonia
```
Dvs. at den fortsetter fra North_Macedonia-lenken


Tar disse to tingene sammen og refaktorerer #1 :+1:

**Reply by eivindkopperud** [reply]
> Jeg tror også at det går helt fint an å komme unna å laste ned alt innhold med data for hver eneste request for å finne lenkene. Ved å legge til ids i rvprop, så kan vi få en revisjonsid for en gitt artikkel:

Bummer. Når man henter ut lenker for en gitt revisjonsid, så får man kun for den nyeste pagen.
Eksempelvis har vi [lenker for nyeste Erna Solberg-artikkel](https://en.wikipedia.org/w/api.php?action=query&format=json&prop=links&revids=1045339806&pllimit=500) og [lenker for eldste Erna Solberg-artikkel](https://en.wikipedia.org/w/api.php?action=query&format=json&prop=links&revids=4282991&pllimit=500) (2004).

Responsene her er identiske. Det [her ](https://en.wikipedia.org/w/api.php?action=query&format=json&prop=revisions&rvlimit=max&titles=Erna_Solberg&rvprop=timestamp|ids|content&rvstartid=4282991)er derimot content'et til Erna Solberg i 2004:
Man kan se at de eneste linkene var "Norway", "politician" og "høyre" 🙃 🔫 

Dvs. at vi er nødt til å parse content likevel, siden det kun er Revision API'et vi kan få lenkene for revisjonene fra. Og da må man ha det på content-format 😭

**Reply by TheodorRene** [reply]
Hvorfor er responsene identiske? Hva skal det endepunktet være godt for da? Det er jo kjedelig.

Så til slutt må du hente ut alle revisjons-id-ene, så gjøre et kall for hver enkelt og så parse lenker inni content?

**Reply by eivindkopperud** [reply]
Vi kan hente ut content, timestamp og revisjonsid vha. én request. Da får vi 50 treff pr. page. Erna har 88 revisjoner i tidspunktet vi ser på, så vi får da 2 requests totalt for Erna Solberg.
